### PR TITLE
feat(vm): support for non-default CloudInit interface and CloudInit storage change

### LINKED ISSUE #485: Control over the CloudInit interface
I am using templates which include a CloudInit image on their `ide0` interface. Because of this, trying to use an `initialization` block while cloning the templates fails.

I would like to be able to specify which IDE interface should be used for the CloudInit image, or to have the provider detect that it isn't on the default interface and none is specified. It should be possible to change the interface without having to recreate the VM.

I have implemented this feature and will submit a PR for it shortly.

### PR DESCRIPTION
**Pull request by tseeker**
### Contributor's Note
Please mark the following items with an [x] if they apply to your PR.
Leave the [ ] if they are not applicable, or if you have not completed the item.
- [x] I have added / updated documentation in `/docs` for any user-facing features or additions.
- [x] I have added / updated templates in `/examples` for any new or updated resources / data sources.
- [x] I have ran `make examples` to verify that the change works as expected. 

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #485

### CODE REVIEW

**Comment by tseeker** [comment]
I was unable to run the examples on a PVE Vagrant box, including without my changes applied (the Ubuntu box boots then stops at the `journald` initialization). I did however test a variety of cloud-init-related configurations on my actual test cluster.

File: proxmoxtf/resource/vm.go [hunk 1]
```diff
@@ -777,9 +780,18 @@ func VM() *schema.Resource {
...
-							ForceNew:    true,
 							Default:     dvResourceVirtualEnvironmentVMInitializationDatastoreID,
 						},
+						mkResourceVirtualEnvironmentVMInitializationInterface: {
+							Type:             schema.TypeString,
+							Description:      "The IDE interface on which the CloudInit drive will be added",
+							Optional:         true,
+							Default:          dvResourceVirtualEnvironmentVMInitializationInterface,
+							ValidateDiagFunc: validator.IDEInterface(),
+							DiffSuppressFunc: func(k, oldValue, newValue string, d *schema.ResourceData) bool {
```
Comment (bpg-dev) [inline proxmoxtf/resource/vm.go:791]: Just curious what use case prompted adding this suppression?

File: proxmoxtf/resource/vm.go [hunk 2]
```diff
@@ -1502,6 +1514,78 @@ func vmCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.D
...
+func findExistingCloudInitDrive(vmConfig *vms.GetResponseData, vmID int, defaultValue string) string {
+	devices := []*vms.CustomStorageDevice{
+		vmConfig.IDEDevice0, vmConfig.IDEDevice1, vmConfig.IDEDevice2, vmConfig.IDEDevice3,
+	}
+	for i, device := range devices {
+		if device != nil && device.Enabled && device.Media != nil && *device.Media == "cdrom" && strings.Contains(
+			device.FileVolume,
+			fmt.Sprintf("vm-%d-cloudinit", vmID),
+		) {
+			return fmt.Sprintf("ide%d", i)
```
Comment (bpg-dev) [inline proxmoxtf/resource/vm.go:1527]: Technically, `device.Interface` _should_ return the proper value, but I also see the point in hardcoding "ideN" in here and not to deal with potential `nil`s.

**Review by bpg — APPROVED** [review]
Awesome work @tseeker, thanks a lot for the PR!!
I also like your refactoring and cleanups. 🚀 

I truly appreciate if you could list test cases you've run the code through? I'm going to start adding more of proper acceptance tests to the provider, instead of relying on static examples, so documenting some corner cases we need to support will be a great help.

**Comment by bpg** [comment]
Also relates to #429

File: proxmoxtf/resource/vm.go (hunk 1 above)
Comment (tseeker) [reply proxmoxtf/resource/vm.go:791]: The suppression prevents the provider from trying to update the resource if it has been initialized with no interface specified. When re-applying, the resource definition will still have an empty `interface` value, but the state (and the Proxmox data) will indicate some `ideX` interface. Without the suppress, it would be detected as changing from `""` to `"ideX"` every time.

File: proxmoxtf/resource/vm.go (hunk 2 above)
Comment (tseeker) [reply proxmoxtf/resource/vm.go:1527]: I didn't know about `device.Interface` ! I would have used that if I had. However your point about potential `nil`s in data which comes from Proxmox and on which the provider has no control is very valid.

Your pointing me at this chunk of code let me see a mistake in the comment for this function though. I will update this shortly.

**Comment by tseeker** [comment]
> I truly appreciate if you could list test cases you've run the code through? I'm going to start adding more of proper acceptance tests to the provider, instead of relying on static examples, so documenting some corner cases we need to support will be a great help.

I tested the following cases : 
  * Creating a VM from a template
    * Template has no CloudInit image
      * With `interface` -> the new VM has a CloudInit image on the specified interface
      * Without `interface` -> the new VM has a CloudInit image on `ide2`
    * Template has a CloudInit image
      * Without `interface` -> the new VM keeps its CloudInit image on the interface it was on
      * With `interface` identical to what was in the template -> the new VM keeps its CloudInit image on the interface it was on
      * With a different `interface` -> the CloudInit image is created on the specified interface
  * Updating a VM
    * Changing the CloudInit `interface`
      * Was `ideX`, removed -> no change on apply
      * Was missing, set to its actual value -> no change on apply
      * Was missing, set to a different interface -> interface removed / re-added on apply
      * Was present, set to a different interface -> interface removed / re-added on apply
    * Changing the `datastore_id` -> CloudInit image is re-created on apply

I didn't test creating a custom VM because the change there is minimal.

**Review by bpg — APPROVED** [review]
